.github/workflows/pr.yml

```diff
@@ -2,27 +2,86 @@ name: Pull Request
 
 on: 
   workflow_dispatch: {}
+  pull_request: {}
+  push:
+    branches:
+      - main
 
 jobs:
   build:
     strategy:
-      matrix:
-        javaVersion: [8, 11, 16, 17]
-        os: [ubuntu-latest, windows-latest]
       fail-fast: false
+      matrix:
+        os: [ubuntu-latest]
+        versions: [
+            { java: 8, gradle: 6.8 },
+            { java: 11, gradle: 6.8 },
+            { java: 16, gradle: 7.3.1 },
+            { java: 17, gradle: 7.3.1 }
+        ]
     runs-on: ${{ matrix.os }}
     steps:
     - uses: actions/checkout@v4
 
     - uses: actions/setup-java@v3
       with:
         distribution: temurin
-        java-version: ${{ matrix.javaVersion}}
+        java-version: ${{ matrix.versions.java}}
         
     - name: Setup Gradle
       uses: gradle/gradle-build-action@v2
       with:
-        gradle-version: 6.8
+        gradle-version: ${{ matrix.versions.gradle }}
+
+    - name: Setup gradle.properties
+      if: matrix.versions.java == 16 || matrix.versions.java == 17
+      run: |
+        echo "org.gradle.jvmargs=-Xmx4096m --add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED --add-exports=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED" >> gradle.properties
     
     - name: Execute Gradle build
-      run: ./gradlew build
\ No newline at end of file
+      run: gradle build
+
+    - name: Build jar
+      run: gradle jar
+
+    - name: Compile integration tests
+      run: gradle compileIntegrationTestJava
+  
+  integration-test:
+    runs-on: ubuntu-latest
+    needs: build
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest]
+        versions: [
+            { java: 8, gradle: 6.8 },
+            { java: 17, gradle: 7.3.1 }
```

Comment:
Optional: Can we add java versions 18, 19, and 20.

Reply:
Let's handle this in a follow-up.